pkg/io/bytestream.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2021 Andy Dalton
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package io
+
+import (
+	"log"
+	"sync"
+)
+
+// DefaultMaxBufferSize is the default maximum number of bytes that a ByteStream
+// will read from the underlying buffer at at once.
+const DefaultMaxBufferSize = 1024
+
+// ByteStream enables clients to stream bytes from an OutputBuffer.
+type ByteStream struct {
+	buffer      OutputBuffer
+	channel     chan []byte
+	maxReadSize int
+
+	// mutex guards the fields that follow
+	mutex            sync.Mutex
+	goroutineStarted bool
+}
+
+// NewByteStream creates and returns a new ByteStream associated with the
+// given buffer with a read buffer size of DefaultMaxBufferSize.
+func NewByteStream(buffer OutputBuffer) *ByteStream {
+	return NewByteStreamDetailed(buffer, DefaultMaxBufferSize)
+}
+
+// NewByteStream creates and returns a new ByteStream associated with the
+// given buffer with a read buffer size of the given maxReadSize.
+func NewByteStreamDetailed(buffer OutputBuffer, maxReadSize int) *ByteStream {
+	return &ByteStream{
+		channel:     make(chan []byte),
+		buffer:      buffer,
+		maxReadSize: maxReadSize,
+	}
+}
+
+// Stream returns a chanel that streams the content of the underlying OutputBuffer.
+func (b *ByteStream) Stream() <-chan []byte {
+
+	bailEarly := true
+
+	func() {
+		b.mutex.Lock()
+		defer b.mutex.Unlock()
+
+		if !b.goroutineStarted {
+			bailEarly = false
+			b.goroutineStarted = true
+		}
+	}()
+
+	// If some other call to Stream has already created the goroutine, then
+	// there's nothing for this call to do other than to return the channel.
+	if bailEarly {
+		return b.channel
+	}
+
+	go func() {
+		var nextByte int64
+		readBuffer := make([]byte, b.maxReadSize)
+
+		for {
+			bufferSize, _ := b.buffer.waitForChange(nextByte)
+
+			// At this point the underlying buffer could be closed, there could
+			// be new bytes in the buffer to process, or both.
+
+			if bufferSize == nextByte {
```

Comment:
Why ignore `closed` here?
```suggestion
			bufferSize, closed := b.buffer.waitForChange(nextByte)

			// At this point the underlying buffer could be closed, there could
			// be new bytes in the buffer to process, or both.

			if bufferSize == nextByte  && closed {
```

Reply:
`waitForChange` returns with `bufferSize == nextByte` only when closed is true, so I think this will be functionally equivalent.  However, I think what you've suggested is easier to read because it depends less on intimate familiarity with the implementation details of `waitForChange`.  I'll make this change -- thank you!